src/ArrayMerger.php

```diff
@@ -160,14 +162,20 @@ protected static function mergeRecursively($default, $precedence, $flags)
             return $precedence; // overwrite default by precedence
         }
         if (! ($flags & self::FLAG_ALLOW_SCALAR_TO_ARRAY_CONVERSION)) {
-            throw new \UnexpectedValueException('different dimensions');
+            if (is_array($default)) {
+                $reason = 'default is an array while precedence is a scalar';
+            }
+            else {
+                $reason = 'default is a scalar while precedence is an array';
```

Comment:
```suggestion
                $reason = '`$default` is a scalar while `$precedence` is an array';
```

Reply:
I've taken that on board, but just quoted 'default', as the value that's caused a problem isn't actually the passed variable $default but a value inside that array.